### DISCUSSION TITLE
Where to post OS API bugs?

### DISCUSSION THREAD

**Discussion by andrewleader** [microsoft/WindowsAppSDK#1876] [Feedback wanted]
We've been discussing the topic of where OS API bugs should be posted... for example, where should developers report WinRT OS APIs that are callable from desktop apps but have a bug? Or Win32 APIs that have a bug? We've seen bugs like https://github.com/microsoft/WindowsAppSDK/issues/1874 and https://github.com/microsoft/WindowsAppSDK/issues/1852 that are cases of this.

We have a few options...

* **Option 1:** The _WindowsAppSDK_ repo is for WinAppSDK + OS API bugs that affect desktop apps
  * _Pros:_ Can easily share/discuss bugs that affect desktop apps... which might be resolved by new WinAppSDK APIs anyways
  * _Cons:_ WinAppSDK repo becomes inflated with bugs that technically aren't its bugs (and that ship on a separate ship cycle)
* **Option 2:** The _WindowsAppSDK_ repo is for WinAppSDK bugs only, Feedback Hub for OS bugs
  * _Pros:_ WinAppSDK repo stays focused on only WinAppSDK
  * _Cons:_ Can be difficult to discuss/discover/report bugs
* **Option 3:** The _WindowsAppSDK_ repo is for WinAppSDK bugs only, we create separate repos for WinRT OS and Win32
  * _Pros:_ WinAppSDK repo stays focused on only WinAppSDK, can still easily share/discuss bugs that affect desktop apps, and those bugs can be captured in their separate ship cycles since it's a separate repo
  * _Cons:_ Multiple repos to search through

**What are people's votes? Vote each option from 1 to 5 stars with 5 stars being the best...** You can copy/paste the table below to share your answer!

```
Options | Vote | Reasoning
--|--|--
Option 1 (all in WinAppSDK): | (1-5 stars) | (Explanation)
Option 2 (WinAppSDK + FB Hub): | (1-5 stars) | (Explanation)
Option 3 (WinAppSDK + repos): | (1-5 stars) | (Explanation) 
```

**Comment by riverar** [comment]
Options | Vote | Reasoning
--|--|--
Option 1 (all in WinAppSDK/OS): | ⭐ | "UWP" encompasses COM, WinRT, and Win32 APIs, some of which are decades old. All of these APIs would then fall into the "OS APIs" bucket and be bug candidates. Filling the repo with OS bugs would make it more difficult to gauge WinAppSDK readiness, find WinAppSDK relevant issues, and I can only imagine, consume large amounts of maintainer time that could be spent on WinAppSDK issues.
Option 2 (WinAppSDK + FB Hub): | ⭐⭐  | Keeping this repository focused on WinAppSDK-specific bugs frees the maintainers from having to triage OS issues. Feedback Hub is an OK alternative, but only if Microsoft commits to actually reading those reports. (It's my understanding the Developer Platform > API Feedback category/sub-category is not currently monitored.) These reports are, however, very difficult to find/share; Feedback Hub also makes use of region-based partitions which could render feedback completely invisible/inaccessible to other developers. There's also no discourse ability. |
Option 3 (WinAppSDK + repos): | ⭐⭐⭐⭐ | Same benefits as Option 2, but with a transparent, developer-focused area to file bugs. Microsoft would need to commit resources to triaging those repositories. (It would be difficult to bring in external folks to assist as well.) Stray issues could easily get transferred to this repository from all repositories under the Microsoft org. |

**Comment by lmadhavan** [comment]
Options | Vote | Reasoning
--|--|--
Option 1 (all in WinAppSDK): | :star::star::star: | This seems like the simplest option given the current state of things. It frees the developer from having to figure out whether an issue is specific to UWP/WinAppSDK or something lower level (WinRT/COM/Win32), which can sometimes be confusing. For example, the official documentation still lists WinRT APIs under UWP even though they can be used in desktop apps.
Option 2 (WinAppSDK + FB Hub): | :star: | I'm not a huge fan of Feedback Hub and I wasn't even aware that there was a section for API issues there, I always thought it was exclusively for end user feedback. I'm sure there are other developers in a similar boat who wouldn't really be looking at Feedback Hub for API issues.
Option 3 (WinAppSDK + repos): | :star::star::star::star::star: | This seems like the ideal long-term solution as it best reflects the actual structure of the products. WinUI and Win2D already have their own repositories and the boundaries of those products are fairly clear. With some effort, the documentation and messaging around the remaining products can be cleaned up and then it should be clear which issues are specific to WinAppSDK and which ones are general OS issues.

**Comment by andrewleader** [comment]
Thanks for the feedback folks! Based on the thumbs up, it looks like nearly everyone is strongly preferring Option 3. We're going to evaluate this in January when more folks return from vacation and see what we can do. No promises about anything happening, but at least we know what the community would prefer, which is the first step toward making progress on this topic! Thank you so much!

**Reply by mrlacey** [reply]
If going with option 3, I have some follow-up questions:

- Who will look at, respond to, and manage these new repos? 
  - If the same people as the WinAppSDK, do they have a preference on this? and suitable tools/workflows for working with multiple repos? (In my experience, when you have lots of repos over which to track issues it's easy for things to be missed.)
  - If not the same people as the WinAppSDK then who? If they are the people who've previously been doing this in private, what is being done to help them work with the community and start using GitHub for these issues?
- Assuming that the code-bases that these OS-level issues relate to will not be made open-source and already have internal/private issue tracking systems, how will details on the public repos be kept up to date and in line with what's tracked privately?
- Will known, existing, issues (from private tracking systems) also be added to the new public repos?
- Is it not better to have issues in both the WinAppSDK repo **and** the other repo? Consider this scenario: You find an issue and log it in the WinAppSDK repo. It is determined that the issue is in an OS component and then moved to a more targeted repo. Then someone else, also using the WinAppSDK, encounters the same problem, and searching the WinAppSDK repo finds no record of the issue and so creates a new issue that is actually a duplicate of the issue that is now in the other repo. Of course, this can be flagged as a duplicate and closed but that's frustrating and demoralizing for the person opening the second/duplicate issue. My preference is to keep an issue in the WinAppSDK repo and link to another issue in the other repo. The issues in each repo can have details (and workarounds?) relevant to the consumers of the product of each repo. [There are other potential benefits of having multiple, linked, issues across repos too.] But, if you don't do this:
  - How will developers know which repos to check for a possible existing record of an issue? (Or are you just going to rely on them doing a general search of the whole web?)
  - Is it possible to offer a tool that searches all the relevant WinDev specific repos?
  - What can be done to avoid giving the impression that by moving an issue out of the WinAppSDK repo it's another team's problem? (Even though it is manifest by using an API/tool that is part of the WinAppSDK.)

**Reply by andrewleader** [reply]
We don't know the answers to all your questions yet, however for your last question as for moving vs creating another issue is a good discussion topic. I personally don't like how moved issues behave (it can be confusing to not be able to find your issue in the repo you created it in, etc). And the problems you bring up are valid concerns too. We'll discuss that topic and see what other teams and the GitHub team thinks on what the best practice there is.

Answering your specific questions...

* How will developers know which repos to check for a possible existing record of an issue? (Or are you just going to rely on them doing a general search of the whole web?)
  * I haven't validated this, but I think most devs do a Google search rather than search on a specific GitHub (which therefore would bring up results from any repo). After all, that would include helpful StackOverflows, etc. And I think devs that know to go to specific GitHub repos likely know enough of the details to infer which repo is the correct repo. Happy to hear your feedback on that though, and that's something we could validate by talking with some more devs!
* Is it possible to offer a tool that searches all the relevant WinDev specific repos?
  * We don't have any plans yet, but I'd doubt we would do that. We could use ZenHub to help aggregate across repos though. We can evaluate whether it becomes a problem or whether devs simply use Google to find issues.
* What can be done to avoid giving the impression that by moving an issue out of the WinAppSDK repo it's another team's problem? (Even though it is manifest by using an API/tool that is part of the WinAppSDK.)
  * We do this today with CsWinRT. We haven't heard feedback that devs got a negative impression from this, have you felt that way?

Overall, are you expressing that you'd rather prefer Option 1? What are your votes on all the options?

**Reply by riverar** [reply]
I found "hey this is the wrong place, we're going to move this to X" works well, no negative feedback yet. Folks are used to this concept from dealings with other support channels. Closing issues and telling them go to another Microsoft repo is terrible and lazy on the maintainer's part.

> Is it not better to have issues in both the WinAppSDK repo and the other repo? Consider this scenario: You find an issue and log it in the WinAppSDK repo. It is determined that the issue is in an OS component and then moved to a more targeted repo. Then someone else, also using the WinAppSDK, encounters the same problem, and searching the WinAppSDK repo finds no record of the issue and so creates a new issue that is actually a duplicate of the issue that is now in the other repo. Of course, this can be flagged as a duplicate and closed but that's frustrating and demoralizing for the person opening the second/duplicate issue.

I think best/common practice is to tell the user the issue is in the wrong place, transfer/close it, and move on. Users can use the search facilities to find relevant issues. Trying to cover every potential search scenario will be untenable and I think the proposed solution just creates a big mess and violates keep it simple principles.

> Is it possible to offer a tool that searches all the relevant WinDev specific repos?

GitHub search already provides this function. (e.g. ["windows error"](https://github.com/search?q=windows+error+repo%3Amicrosoft%2Fwindowsappsdk&type=issues))

**Reply by mrlacey** [reply]
I guess I've learned a lesson here: Don't create an issue in a repo unless I'm certain it's in the right repo, and check the whole internet/GitHub for other reports of an issue first too.

**Reply by andrewleader** [reply]
@mrlacey I don't think that's the takeaway... if you create an issue in the wrong repo, teams that manage the repo will simply inform you and either move the issue for you, or close it and tell you where you should post it. It's not the end of the world and it's okay if you or someone else accidentally creates an issue in the "wrong" repo :) but to get your issues addressed fastest, it'd be best if you create them in the right repo after learning the right repo.

**Reply by mrlacey** [reply]
But creating a good repro takes ages! I have created accidental duplicate issues in the past and wasted hours in the process. I'm not saying everyone should do the same, it was more a note for myself as much as anything else.

**Reply by michael-hawker** [reply]
I definitely would lean towards 1. Having everything related to Windows APIs in one place makes it easier for folks to find existing issues. It's also unclear if something is related to the platform or a specific SDK when working on things. And if there's new stuff coming or should be used, folks here would know that the best.

This is where issue templates and auto-labeling could be handy to help differentiate them in the case folks know ahead of time what part of the platform an issue pertains to.

**Reply by micolous** [reply]
I saw [this discussion was referenced for a platform bug][0], but while it looks like consensus was achieved some time ago, was anything finalised?

There are also [APIs like WebAuthn, which already had a public repository][1] (which would align with "Option 3"), but it doesn't presently accept issues or discussions. There's a contact email address, but _unlike_ the discussions here, that's all 1:1, so its users can't even help each other out.

Thanks!

[0]: https://github.com/microsoft/win32metadata/issues/1607#issuecomment-1627407822
[1]: https://github.com/microsoft/webauthn

**Comment by mdtauk** [comment]
OS API bugs/requests/asks should have a public forum for upvoting and triaging.  Like the new Product board that was added.

A place where the users/developers/community can help with prioritisation.

**Comment by jaigak** [comment]
The real problem with Feedback Hub is that it is too limited. You can't edit feedback, you can't use rich formatting and you can't see feedback from all countries and languages (only feedback from your own country and language, as well as English).

**Reply by DrusTheAxe** [reply]
Good feedback for FeedbackHub :-) @RDMacLachlan can you pass along word?

**Reply by riverar** [reply]
It's by design don't waste your time. And feedback hub is regularly wiped per policy making it a terrible location for OS bugs.

**Reply by DarranRowe** [reply]
When it comes to API feedback, I have gotten into the habit of just linking to a GitHub repo. That takes care of the vast majority of the deficiencies since the formatting is taken care of by GitHub itself. It is also useful to communicate the ever so useful minimal reproduction sample too. The README.md file can also be used to add any of the extra details that any character limit may inhibit.

This makes me wonder about me looking for ways to game the system.

**Comment by OrbintSoft** [comment]
So in the end what is the correct place to report windows OS API bugs? I tried to report this issue here https://github.com/microsoft/WindowsAppSDK/issues/4955, but they closed saying I reported in the wrong repository.

I can report it Feedback Hub, but it will never get attention from developer, it's a good tool for an average user to give windows feedback, but  it's not a serious tool report bugs in the API, report source code and start a constructive technical discussion.

**Reply by DrusTheAxe** [reply]
@RDMacLachlan what's the current recommendation?

**Reply by RDMacLachlan** [reply]
From my experience, and conversations with OS teams, Feedback Hub is the recommended channel for reporting OS-level bugs. The relevant OS teams do actively monitor Feedback Hub submissions and triage them regularly.

**Reply by OrbintSoft** [reply]
There are like millions of feedbacks on windows feeback. You want me let me believe that there is some core OS engineer that is analyzing every issue and triaging them? Personally I think there are just some people that are looking to the issues with most upvotes , they check if easy to reproduce by looking a screenshots and steps in the description and only if they reproduce they send the details to the engineer. That is good tool for regular windows users where they report feedbacks and bug from daily user experience . But if I. Report as developer a very technical issue, with c++ code, hard to read because, no code formatting, with a very technical explanation and a link to github / stackoverflow. First very few people will upvote, because most people that use feedback hub will not understand . Second I bet no one at Microsoft will ever even try to analyze and reproduce. Prove me if I am wrong https://aka.ms/AAwzatc

**Comment by StefanoBalzarottiNozomi** [comment]
Option 1: No, you close the issue and mark it as not planned.
Option 2: It is impossible to provide a technical detail/report of the issue, the feedback are flooded and most of the times never read.
Option 3: There is no option 3.

The only way to report a bug in OS API is to buy a support plan.

**Reply by LaoshuBaby** [reply]
This is M$
